alts/src/main/java/io/grpc/alts/GoogleDefaultChannelBuilder.java

```diff
@@ -0,0 +1,237 @@
+import io.grpc.netty.InternalNettyChannelBuilder.TransportCreationParamsFilterFactory;
+import io.grpc.netty.NettyChannelBuilder;
+import io.netty.handler.ssl.SslContext;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLException;
+
+/**
+ * Google default version of {@code ManagedChannelBuilder}. This class sets up a secure channel
+ * using ALTS if applicable and using TLS as fallback.
+ */
+public final class GoogleDefaultChannelBuilder
+    extends ForwardingChannelBuilder<GoogleDefaultChannelBuilder> {
+
+  private final NettyChannelBuilder delegate;
+  private TcpfFactory tcpfFactoryForTest;
+
+  /** "Overrides" the static method in {@link ManagedChannelBuilder}. */
+  public static final GoogleDefaultChannelBuilder forTarget(String target) {
+    return new GoogleDefaultChannelBuilder(target);
+  }
+
+  /** "Overrides" the static method in {@link ManagedChannelBuilder}. */
+  public static GoogleDefaultChannelBuilder forAddress(String name, int port) {
+    return forTarget(GrpcUtil.authorityFromHostAndPort(name, port));
+  }
+
+  private GoogleDefaultChannelBuilder(String target) {
+    delegate =
+        NettyChannelBuilder.forTarget(target)
+            .keepAliveTime(20, TimeUnit.SECONDS)
+            .keepAliveTimeout(10, TimeUnit.SECONDS)
+            .keepAliveWithoutCalls(true);
+  }
+
+  @Override
+  protected NettyChannelBuilder delegate() {
+    return delegate;
+  }
+
+  @Override
+  public ManagedChannel build() {
+    CheckGcpEnvironment.check(/*enableUntrustedAlts=*/ false);
+    TcpfFactory tcpfFactory = new TcpfFactory();
+    InternalNettyChannelBuilder.setDynamicTransportParamsFactory(delegate(), tcpfFactory);
+    tcpfFactoryForTest = tcpfFactory;
+    try {
+      return new GoogleDefaultChannel(
+          delegate().build(), GoogleCredentials.getApplicationDefault());
+    } catch (IOException e) {
+      String message = "Failed to obtain default credentials: " + e.getMessage();
+      throw new RuntimeException(message, e);
+    }
+  }
+
+  @VisibleForTesting
+  @Nullable
+  TransportCreationParamsFilterFactory getTcpfFactoryForTest() {
+    return tcpfFactoryForTest;
+  }
+
+  private static final class TcpfFactory implements TransportCreationParamsFilterFactory {
+    private final SslContext sslContext;
+    private final AltsClientOptions handshakerOptions =
+        new AltsClientOptions.Builder()
+            .setRpcProtocolVersions(RpcProtocolVersionsUtil.getRpcProtocolVersions())
+            .build();
+
+    private final TsiHandshakerFactory altsHandshakerFactory =
+        new TsiHandshakerFactory() {
+          @Override
+          public TsiHandshaker newHandshaker() {
+            // Used the shared grpc channel to connecting to the ALTS handshaker service.
+            ManagedChannel channel = HandshakerServiceChannel.get();
```

Comment:
I've created  https://github.com/grpc/grpc-java/issues/4755. Let fix this quarter in a separate PR. It would be fairly easy to unref the channel if nobody is using it. I like to keep the thread open until application closes. Otherwise creating/closing thread could also be expensive. Is there concept of fiber (i.e., lightweight thread) in Java?

Reply:
> I like to keep the thread open until application closes.

No. The threads are actually the primary problem.

> Otherwise keep creating/closing thread could also be expensive.

This is what our SharedResourceHolder deals with, to a degree. It waits a second before actually freeing, in case there will be another use soon. In general though, we would expect users to have a long-lived Channel, so there shouldn't be rampant thread creation.

> Is there concept of fiber (i.e., lightweight thread) in Java?

Nope. I wish.

Reply:
Let me take a look at SharedResourceHolder. I may schedule a mtg with you and @carl-mastrangelo. I remember old time ago, @carl-mastrangelo  and I thought about SharedResourceHolder, but we did not use it. I don't remember the exact reason.